<a href="https://colab.research.google.com/github/amoukrim/AI/blob/main/Week8/ExerciseXP/exercicexp_w_8_d2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Exercice 1 – Étape 1 : Gather Definitions
📌 Objectif de l’étape : Rechercher et noter les définitions des trois niveaux d’ouverture d’un modèle LLM.

* Fully Open : Le code source, l’architecture du modèle, et les poids

préentraînés sont disponibles publiquement et sans restriction majeure.

* Weights Released : Les poids du modèle sont accessibles (souvent pour l’affinage), mais le code source ou les données d'entraînement ne sont pas forcément partagés.

Architecture Only : Seule la structure (layers, dimensions, attention mechanism, etc.) du modèle est publiée, sans les poids préentraînés.

##Exercice 1 – Étape 2 : Identify Key Characteristics

1. Fully Open
✅ Ce qui est ouvert : Code source, poids préentraînés, architecture complète.

⚠️ Ce que tu peux faire : Tout – inspection, modification, réentraînement, déploiement commercial.

2. Weights Released
✅ Ce qui est ouvert : Poids préentraînés uniquement.

⚠️ Ce que tu peux faire : Fine-tuning possible, mais accès limité à la logique interne du code ou aux données d’origine.

3. Architecture Only
✅ Ce qui est ouvert : Description de l’architecture du modèle (schéma, dimensions, couches…).

⚠️ Ce que tu peux faire : Impossible d’utiliser ou d’affiner sans réentraîner les poids à partir de zéro.



##Exercice 1 – Étape 3 : Compare Them Side by Side

| Niveau d’ouverture | Qu’est-ce qui est ouvert ?           | Impact sur l'entraînement / modification                         |
| ------------------ | ------------------------------------ | ---------------------------------------------------------------- |
| Fully Open         | Code + poids + architecture          | Tout est possible : inspection, fine-tuning, déploiement         |
| Weights Released   | Poids uniquement                     | Fine-tuning possible, mais pas de modification de l’architecture |
| Architecture Only  | Structure du modèle (sans les poids) | Réentraînement complet requis ; pas de fine-tuning possible      |


#Exercice 1 – Étape 4 : Rédiger un paragraphe comparatif

Objectif : Résumer les trois niveaux d’ouverture en 3 à 5 phrases, en insistant sur leurs avantages ou limites.

Les niveaux d’ouverture des LLM varient selon les composants accessibles. Un modèle Fully Open met à disposition son code, ses poids préentraînés et son architecture, ce qui permet une transparence totale, une modification complète et un réentraînement sans restriction. Un modèle avec uniquement les Weights Released offre ses poids pour un fine-tuning possible, mais le code source ou l’architecture détaillée peuvent manquer, limitant la personnalisation. Enfin, un modèle avec Architecture Only ne fournit que le schéma du modèle, sans les poids, ce qui empêche tout usage direct ou fine-tuning, à moins de réentraîner depuis zéro. Le degré d’ouverture choisi influence donc fortement ce qu’on peut faire avec le modèle dans un contexte pratique.



##Exercice 1 – Étape 5 : Répondre au cas d’usage en santé

Le niveau Fully Open est essentiel pour ce cas, car il permet un accès complet au modèle pour le fine-tuning sur des données cliniques. Ce niveau garantit une transparence et un contrôle total, ce qui est crucial pour respecter les exigences de confidentialité et de conformité dans le domaine médical.

#Exercise 2: License Check for SaaS Use

##Étape 1 : Sélection de 2 modèles open source sur Hugging Face :

mistralai/Mistral-7B-Instruct

* Élément de liste
* meta-llama/Llama-2-7b-chat-hf

##Étape 2 : Localiser la licence
Objectif : Repérer le type de licence sur chaque page de modèle.


| Modèle                    | Licence affichée sur Hugging Face        |
| ------------------------- | ---------------------------------------- |
| Mistral-7B-Instruct       | Apache 2.0                               |
| Llama-2-7b-chat-hf (Meta) | Meta Llama 2 Community License Agreement |


## Étape 3 : Vérifier l’usage commercial


Mistral (Apache 2.0) :
✔️ Usage commercial autorisé sans condition majeure. Licence permissive.

LLaMA 2 (Meta license) :
⚠️ Usage commercial conditionnel.
Tu dois accepter la licence, respecter certaines limitations (ex : pas pour certains secteurs, pas de redistribution des poids).

## Étape 4 : Identifier les restrictions spécifiques


Mistral (Apache 2.0) :

Aucune restriction particulière

Attribution recommandée mais non obligatoire

Pas de limite utilisateur ou contrainte géographique

LLaMA 2 (Meta license) :

❗️ Pas de modification de la licence

❗️ Attribution obligatoire

❗️ Restrictions sur les cas d’usage sensibles (ex : défense, surveillance)

❗️ Accord requis (clique sur "Agree and Access")

❗️ Pas de distribution des poids modifiés



##

##Étape 5 : Checklist comparative en Markdown

Voici ta checklist complète en format Markdown, prête à copier dans ton livret ou ton dépôt :

- [ ] **mistralai/Mistral-7B-Instruct**  
- Type de licence:  
  - [x] Apache 2.0  
- Usage commercial autorisé:  
  - [x] Oui  
- Restrictions:  
  - [ ] Aucune restriction majeure  
  - [ ] Attribution recommandée, non obligatoire  

- [ ] **meta-llama/Llama-2-7b-chat-hf**  
- Type de licence:  
  - [x] Meta Llama 2 Community License  
- Usage commercial autorisé:  
  - [x] Conditionnel  
- Restrictions:  
  - [x] Acceptation obligatoire de la licence  
  - [x] Attribution obligatoire  
  - [x] Restrictions sur certains domaines d’usage  
  - [x] Interdiction de redistribuer les poids modifiés  


#Exercise 3: LLM Matchmaker Challenge

## Étape 1 : Analyser les besoins des équipes
📌 Objectif : Identifier les contraintes clés de chaque équipe pour choisir un modèle LLM adapté.

#Voici les trois équipes à analyser :

| Équipe         | Besoin principal                                                    |
| -------------- | ------------------------------------------------------------------- |
| **LegalTech**  | Chatbot orienté logique, **inférence uniquement sur CPU**           |
| **EdTech**     | Raisonnement math/logiciel sur **ordinateur portable bas de gamme** |
| **Global NGO** | Doit gérer **5+ langues** efficacement (multilingue)                |

Pour LegalTech → modèles petits, efficaces sur CPU, bons en logique (ex : BoolQ, ReClor).


Pour EdTech → faible empreinte mémoire + bonnes performances en mathématiques (ex : GSM8K).

Pour Global NGO → très bon support multilingue (ex : évalué sur FLORES-200, XGLUE, etc.).


#Étape 2 : Rechercher sur Hugging Face
Objectif : Trouver 3–5 modèles par besoin à l’aide des filtres Hugging Face.

Voici comment filtrer :
Taille ≤ 7B

Tags : “logic”, “math”, “multilingual”

Compatibles CPU ou quantized (4-bit, GGUF)

Sélection de modèles par équipe
🔹 1. LegalTech – Logique + Inférence CPU


| Modèle                            | Taille | Archi       | Avantage clé                                  |
| --------------------------------- | ------ | ----------- | --------------------------------------------- |
| **Nous-Hermes-2 4-bit GGUF**      | \~7B   | Mixtral     | Très rapide en quantifié, très bon en logique |
| **phi-2** (Microsoft)             | 2.7B   | Transformer | Ultra léger, excellent en logique (BoolQ)     |
| **mistralai/Mistral-7B-Instruct** | 7B     | Mistral     | Performant, peut être quantifié pour CPU      |
| **openchat/openchat-3.5-0106**    | 7B     | LLaMA-based | Fort en raisonnement, fonctionne en int4      |

🎯 Meilleurs pour CPU + logique : phi-2, Nous-Hermes-2 (GGUF), Mistral 7B quantifié

 2. EdTech – Math + Ordi portable bas de gamme
 | Modèle                           | Taille | Bench Math  | Notes                                   |
| -------------------------------- | ------ | ----------- | --------------------------------------- |
| **phi-2** (Microsoft)            | 2.7B   | GSM8K \~74% | Ultra léger, très bon en math + logique |
| **TinyLlama-1.1B-Chat**          | 1.1B   | -           | Fonctionne même avec très peu de RAM    |
| **WizardMath-1.5B**              | 1.5B   | GSM8K \~67% | Optimisé pour math, quantifiable        |
| **deepseek-coder-1.3B-instruct** | 1.3B   | -           | Bon pour logique et calcul              |

🎯 Meilleurs pour math + faible RAM : phi-2, WizardMath-1.5B, TinyLlama

3. Global NGO – Multilingue (5+ langues)

| Modèle                           | Taille | FLORES-200 | Langues      | Notes                                        |
| -------------------------------- | ------ | ---------- | ------------ | -------------------------------------------- |
| **BLOOMZ-3B**                    | 3B     | \~64%      | 50+          | Traduit bien, multilingue généraliste        |
| **mBART-50**                     | 1.3B   | \~60%      | 50+          | Très léger, multilingue (traduction surtout) |
| **XGLM-4.5B**                    | 4.5B   | \~70%      | 30+          | Excellent multilingue pour génération        |
| **Mistral 7B Instruct (quant.)** | 7B     | -          | Très partiel | Multilingue limité (plutôt EN/FR)            |



Tableau final – LLM Matchmaker Challenge

| Équipe         | Besoins spécifiques                                      | 🧠 **Votre choix**                                                                                       |
| -------------- | -------------------------------------------------------- | -------------------------------------------------------------------------------------------------------- |
| **LegalTech**  | Modèle rapide orienté logique, **CPU-only**              | `phi-2` (Microsoft) – Petit modèle très performant sur BoolQ (logique) et très léger pour inférence CPU. |
| **EdTech**     | Modèle **math/logic** compatible **laptop bas de gamme** | `WizardMath-1.5B` – Excellent compromis taille/perf en mathématiques (GSM8K), facile à faire tourner.    |
| **Global NGO** | Support de **5+ langues** (multilingue large)            | `BLOOMZ-3B` – Modèle open, multilingue performant, taille raisonnable, FLORES-200 solide.                |

**Justifications rapides**

phi-2 : idéal pour LegalTech car il offre d’excellentes performances logiques tout en étant <3B de paramètres.

WizardMath-1.5B : adapté à EdTech avec très bonne perf GSM8K + faible mémoire nécessaire.

BLOOMZ-3B : conçu pour la génération multilingue, parfait pour un usage global ONG.



## Étape bonus : Résumé des filtres utilisés

- **LegalTech**  
  - Filtres : "logic", CPU, ≤ 7B, quantized  
  - URL : https://huggingface.co/models?search=logic&library=transformers&sort=downloads&quantized=true

- **EdTech**  
  - Filtres : "math", quantized, ≤ 3B  
  - URL : https://huggingface.co/models?search=math&sort=downloads&quantized=true

- **Global NGO**  
  - Filtres : "multilingual", FLORES-200, ≤ 5B  
  - URL : https://huggingface.co/models?search=multilingual&sort=downloads


# Exercice 4 : Local Readiness Audit

##Étape 1 : Récupérer les spécifications système
📌 **Objectif** : Connaître les trois éléments suivants sur ta machine :
🧠 RAM totale

💾 Espace disque libre

💻 Système d’exploitation (Linux ou WSL2 si Windows)

🔧 Selon ton OS, voici comment faire :
▶️ Sur Linux/macOS :
Ouvre un terminal et tape :
free -h         # Pour voir la RAM
df -h           # Pour voir l’espace disque
uname -r        # Pour la version du noyau Linux

▶️ Sur Windows avec WSL2 :
Ouvre PowerShell ou un terminal WSL.

Tape :
wsl --status     # Pour voir si WSL2 est actif
Et pour le reste, utilise le Gestionnaire des tâches (onglet Performance) pour RAM et disque.

#Étape 2 : Remplir le tableau d’audit

| Requirement               | Your System Specs    | Meets Requirement? |
| ------------------------- | -------------------- | ------------------ |
| RAM (≥ 16 GB)             | 8 GB                 | ❌                  |
| Free Disk Space (≥ 40 GB) | 100 GB               | ✅                  |
| OS (Linux/WSL2)           | Windows 11 (non-WSL) | ❌                  |


#✅ Étape 3 : Analyse rapide

🔴 RAM insuffisante : 8 GB est trop juste pour faire tourner un modèle 7B, même quantifié. Un modèle 2–3B quantifié serait plus réaliste.

🟢 Espace disque OK : tu es largement au-dessus des 40 GB minimum.

🔴 OS non prêt : sous Windows sans WSL2 ou Linux, llama.cpp ou text-generation-webui fonctionneront mal voire pas du tout.

##✅ Étape 4 : Vérifier la compatibilité llama.cpp

| Problème           | Solution recommandée                                   |
| ------------------ | ------------------------------------------------------ |
| RAM insuffisante   | 👉 Utiliser des modèles ≤ 3B quantifiés (int4 ou GGUF) |
| OS non compatible  | 👉 Installer **WSL2** + Ubuntu sous Windows            |
| CPU pas compatible | 👉 Utiliser un cloud léger (Colab, RunPod…)            |


##Étape 5 : Conseils d’amélioration

| Problème           | Solution recommandée                                   |
| ------------------ | ------------------------------------------------------ |
| RAM insuffisante   | 👉 Utiliser des modèles ≤ 3B quantifiés (int4 ou GGUF) |
| OS non compatible  | 👉 Installer **WSL2** + Ubuntu sous Windows            |
| CPU pas compatible | 👉 Utiliser un cloud léger (Colab, RunPod…)            |




🎯 **Conclusion :**

Ma machine n’est pas prête à faire tourner un modèle 7B localement, mais je pourrais utiliser des modèles plus petits (TinyLlama, phi-2, 1.3B quantifiés) si tu installes WSL2 ou utilises un outil comme Oobabooga avec GGUF 4-bit.

##Étape complémentaire : Vérifier le support AVX2
Le Intel i7-8550U :

Architecture : Kaby Lake Refresh (8e génération)

Fréquence de base : 1.80GHz (boost jusqu’à ~4.0GHz)

Supporte AVX2 ✅

Ne supporte pas AVX-512 ❌ (ce n’est pas nécessaire pour llama.cpp)



**Conclusion technique :**

| Élément                   | Résultat                                      |
| ------------------------- | --------------------------------------------- |
| Instruction AVX2          | ✅ Supporté                                    |
| Compatible avec llama.cpp | ✅ Oui, pour modèles optimisés AVX2            |
| Utilisation recommandée   | ✅ GGUF 4-bit, modèles ≤ 3B (phi-2, TinyLlama) |


**Résumé final du Local Readiness Audit (mis à jour)**

| Requirement               | Your System Specs        | Meets Requirement? |
| ------------------------- | ------------------------ | ------------------ |
| RAM (≥ 16 GB)             | 8 GB                     | ❌                  |
| Free Disk Space (≥ 40 GB) | 100 GB                   | ✅                  |
| OS (Linux/WSL2)           | Windows 11 (non-WSL)     | ❌                  |
| AVX support (AVX2 min)    | Intel i7-8550U (AVX2 OK) | ✅                  |


**Recommandations** :
✅ Ton CPU est compatible avec llama.cpp.

❌ Tu dois activer WSL2 + Ubuntu pour un environnement Linux léger sous Windows.

❌ 8 GB de RAM est trop juste pour un 7B, vise ≤ 3B quantifié (4-bit).

#Exercice 5 : Benchmark-Based Model Explorer

#Tableau comparatif : Benchmark-Based Model


| Model Name                          | HellaSwag Score | MMLU Score | License Type                          | Ideal Use Case                                 |
| ----------------------------------- | --------------- | ---------- | ------------------------------------- | ---------------------------------------------- |
| **mistralai/Mixtral-8x7B-Instruct** | **85.2**        | 74.5       | Apache 2.0                            | Assistant général haut niveau (équilibré)      |
| **openchat/openchat-3.5-0106**      | **88.3**        | 65.2       | MIT                                   | Chatbot conversationnel logique (raisonnement) |
| **meta-llama/Llama-3-8B-Instruct**  | 81.0            | **78.7**   | Meta Llama 3 License (non-permissive) | Tuteur académique multi-matière                |


** Interprétation :**
🔵 Mixtral-8x7B : Excellente performance équilibrée. Idéal pour assistants polyvalents ou SaaS multitâche.

🔴 OpenChat-3.5 : Ultra performant en logique/réponse à choix multiples (HellaSwag), moins bon en connaissances spécialisées.

🟢 LLaMA 3-8B : Top pour la compréhension académique (MMLU), mais attention : licence restreinte, usage commercial limité.

#Étape 3 : Résumé de la sélection

**Modèle A (équilibré)**  
- Nom : mistralai/Mixtral-8x7B-Instruct  
- HellaSwag : 85.2  
- MMLU : 74.5  
- Licence : Apache 2.0  
- Cas d’usage : Assistant général de haut niveau

**Modèle B (raisonnement)**  
- Nom : openchat/openchat-3.5-0106  
- HellaSwag : 88.3  
- MMLU : 65.2  
- Licence : MIT  
- Cas d’usage : Chatbot logique, QCM, interface conversationnelle

**Modèle C (académique)**  
- Nom : meta-llama/Llama-3-8B-Instruct  
- HellaSwag : 81.0  
- MMLU : 78.7  
- Licence : Meta Llama 3 Community License  
- Cas d’usage : Système de tutorat ou assistant pédagogique spécialisé

**Réflexion courte** :

Choisir un LLM ne devrait pas se baser uniquement sur la notoriété, mais sur les benchmarks objectifs. Les scores HellaSwag et MMLU permettent de choisir un modèle adapté à des cas concrets : raisonnement, connaissances générales ou usages polyvalents.


# Exercice 6 : Cloud vs Local Deployment Plan

## Étape 1 : Comparer les deux environnements

**Distinguer les avantages et inconvénients entre :**

Local Deployment (exécuter un modèle sur ton PC ou serveur)

Cloud Deployment (utiliser Colab, RunPod, Hugging Face Inference, etc.)


#Étape 2 : Liste des Pour / Contre (5 bullet points)

Voici une comparaison équilibrée que tu peux reprendre ou adapter :

🖥️ Local Deployment
✔️ Faible latence, pas de cold start (le modèle est toujours dispo)

❌ Consomme beaucoup de RAM et CPU/GPU local

✔️ Contrôle total sur les données (pas d’envoi à des serveurs tiers)

❌ Configuration initiale complexe (installations, compilation, compatibilité)

❌ Limité par ton matériel (taille de modèle, vitesse d'inférence)

☁️ Cloud Deployment
✔️ Facile à démarrer (Google Colab, RunPod, Hugging Face Spaces)

✔️ Accès rapide à des GPUs puissants sans investissement

❌ Latence potentielle (temps de démarrage, files d'attente)

❌ Risques de confidentialité (données transitant par le cloud)

❌ Coût à long terme (abonnements, crédits GPU, limites horaires)

##Étape 3 (Optionnelle) : Test Colab – modèle 7B

In [3]:
from huggingface_hub import login
login()  # Colle ici ton token Hugging Face


In [4]:
!pip install transformers accelerate

from transformers import AutoModelForCausalLM, AutoTokenizer
import torch, time

model_name = "meta-llama/Llama-2-7b-chat-hf"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", torch_dtype=torch.float16)

inputs = tokenizer("Hello, world!", return_tensors="pt").to(model.device)
start = time.time()
outputs = model.generate(**inputs, max_new_tokens=50)
print(tokenizer.decode(outputs[0]))
print("Elapsed:", time.time() - start, "seconds")


tokenizer_config.json:   0%|          | 0.00/1.62k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.84M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/414 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/614 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/26.8k [00:00<?, ?B/s]

Fetching 2 files:   0%|          | 0/2 [00:00<?, ?it/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.98G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/3.50G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/188 [00:00<?, ?B/s]

<s> Hello, world!
I'm so excited to finally share my new website with you! This has been a labor of love for me, and I'm thrilled to finally have it up and running.

As you can see, the website is all
Elapsed: 4.588147401809692 seconds


#Exercice 6 : Cloud vs Local Deployment Plan.

##Résultat de ton test Colab
Modèle utilisé : meta-llama/Llama-2-7b-chat-hf

Temps de génération : ≈ 4.6 secondes

✅ Le modèle a généré une réponse fluide et réaliste (type assistant conversationnel)



#✅ Récapitulatif de ton plan de déploiement

**Cloud vs Local Deployment – Comparatif final**

| Aspect              | Local Deployment                          | Cloud Deployment (Colab)                  |
| ------------------- | ----------------------------------------- | ----------------------------------------- |
| **Latence**         | ✅ Très faible (aucun démarrage à froid)   | ❌ Peut varier selon charge ou inactivité  |
| **Ressources**      | ❌ Limitée à 8 GB RAM (modèles < 3B)       | ✅ Accès à GPU, même pour 7B+ modèles      |
| **Confidentialité** | ✅ Données en local                        | ❌ Données passent par internet/cloud      |
| **Simplicité**      | ❌ Complexe à configurer (WSL2, llama.cpp) | ✅ Facile à tester via Colab, Hugging Face |
| **Coût**            | ✅ Aucun coût si matériel déjà en place    | ❌ Payant à long terme (abos ou crédits)   |


**Conclusion :**

Le test via Google Colab a permis d’exécuter un modèle LLaMA 2-7B avec succès en moins de 5 secondes, démontrant la puissance du cloud pour des déploiements rapides sans contraintes matérielles locales. En revanche, pour des raisons de confidentialité ou de coûts continus, un déploiement local reste pertinent si l’on utilise des modèles quantifiés plus petits et que l’environnement est prêt.